src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,401 @@
+        : Async<int64 * ResolvedEvent[]> = async {
+        let mergeBatches (batches : AsyncSeq<int64 * ResolvedEvent[]>) = async {
+            let mutable versionFromStream = -1L
+            let! (events : ResolvedEvent[]) =
+                batches
+                |> AsyncSeq.map (fun (reportedVersion, events)-> versionFromStream <- max reportedVersion versionFromStream; events)
+                |> AsyncSeq.concatSeq
+                |> AsyncSeq.toArrayAsync
+            let version = versionFromStream
+            return version, events }
+        let call pos = loggedReadSlice conn streamName batchSize pos
+        let retryingLoggingReadSlice pos = Log.withLoggedRetries retryPolicy "readAttempt" (call pos)
+        let log = log |> Log.prop "batchSize" batchSize |> Log.prop "stream" streamName
+        let batches : AsyncSeq<int64 * ResolvedEvent[]> = readBatches log retryingLoggingReadSlice maxPermittedBatchReads startPosition
+        let! t, (version, events) = mergeBatches batches |> Stopwatch.Time
+        log |> logBatchRead streamName t events batchSize version
+        return version, events }
+
+module Token =
+    let create streamVersion : StreamToken =
+        { value = box streamVersion
+          version = streamVersion
+          streamBytes = -1 }
+    let supersedes struct (current, x) =
+            x.version > current.version
+type MessageDbConnection(readConnection, [<O; D(null)>]?writeConnection, [<O; D(null)>]?readRetryPolicy, [<O; D(null)>]?writeRetryPolicy) =
+    member _.ReadConnection = readConnection
+    member _.ReadRetryPolicy = readRetryPolicy
+    member _.WriteConnection = defaultArg writeConnection readConnection
+    member _.WriteRetryPolicy = writeRetryPolicy
+
+type BatchOptions(getBatchSize : Func<int>, [<O; D(null)>]?batchCountLimit) =
+    new (batchSize) = BatchOptions(fun () -> batchSize)
+    member _.BatchSize = getBatchSize.Invoke()
+    member _.MaxBatches = batchCountLimit
+
+[<RequireQualifiedAccess; NoComparison; NoEquality>]
+type GatewaySyncResult = Written of StreamToken | ConflictUnknown
+
+type MessageDbContext(connection : MessageDbConnection, batchOptions : BatchOptions) =
+    let isResolvedEventEventType (tryDecode, predicate) (e : ResolvedEvent) = predicate (tryDecode e.Data)
+    let tryIsResolvedEventEventType predicateOption = predicateOption |> Option.map isResolvedEventEventType
+    new (   connection : MessageDbConnection,
+            // Max number of Events to retrieve in a single batch. Also affects frequency of RollingSnapshots. Default: 500.
+            [<O; D null>] ?batchSize) =
+        MessageDbContext(connection, BatchOptions(batchSize = defaultArg batchSize 500))
+    member val BatchOptions = batchOptions
+
+    member _.TokenEmpty = Token.create -1L
+    member _.LoadBatched streamName log (tryDecode, isCompactionEventType) : Async<StreamToken * 'event[]> = async {
```

Comment:
I guess you know `isCompactionEventType` is redundant unless you intend to put in the embedded snapshots and/or reset event to be meaningful support (which requires reading backwayrd)